File: 08_Bellman_Form_Algorithm/bellman_ford.py
import sys
INT_MAX = sys.maxsize

class Edge:
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c 

n, e = input().split()
edge_list = []
dist = []

for i in range(int(e)):
    a, b, c = input().split()
    edge_list.append(Edge(a, b, c))

for i in range(int(n)):
    dist.append(INT_MAX)

dist[0] = 0
cycle = False
for i in range(int(n) - 1):
    for ed in edge_list:
        a = int(ed.a)
        b = int(ed.b)
        c = int(ed.c)
        if dist[a] != INT_MAX and dist[a] + c < dist[b]:
            dist[b] = dist[a] + c
    
    for ed in edge_list:
        a = int(ed.a)
        b = int(ed.b)
        c = int(ed.c)
        if dist[a] != INT_MAX and dist[a] + c < dist[b]:
            cycle = True
            break;

if not cycle:
    print("DEBUG::")
    for i in range(int(n)):
        print(f"{i} -> {dist[i]}")
else:
    print("Negative weighted cycle detected")


"""
4 4
0 2 5
0 3 12
2 1 2
1 3 3
"""
"""
4 4
0 1 5
1 2 3
2 3 2
3 1 -6
"""
"""
4 4
0 1 5
0 3 6
1 2 3
2 3 5
"""

File: 08_Bellman_Form_Algorithm/bellman_ford.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

class Edge
{
public:
    int a, b, c;
    Edge(int u, int v, int w)
    {
        a = u;
        b = v;
        c = w;
    }
};

int main()
{
    int n, e;
    cin >> n >> e;

    vector<Edge> edge_list;
    vector<int> dist(n, INT_MAX);

    for (int i = 0; i < e; i++)
    {
        int a, b, c;
        cin >> a >> b >> c;
        edge_list.push_back(Edge(a, b, c));
    }

    // Bellman-Ford Algorithm
    dist[0] = 0;
    bool cycle = false;

    for (int i = 0; i < n - 1; i++)
    {
        for (const auto &ed : edge_list)
        {
            if (dist[ed.a] != INT_MAX && dist[ed.a] + ed.c < dist[ed.b])
            {
                dist[ed.b] = dist[ed.a] + ed.c;
            }
        }
    }

    // Check for negative-weight cycles
    for (const auto &ed : edge_list)
    {
        if (dist[ed.a] != INT_MAX && dist[ed.a] + ed.c < dist[ed.b])
        {
            cycle = true;
            break;
        }
    }

    if (!cycle)
    {
        cout << "DEBUG::\n";
        for (int i = 0; i < n; i++)
        {
            if (dist[i] == INT_MAX)
                cout << i << " -> INF\n";
            else
                cout << i << " -> " << dist[i] << "\n";
        }
    }
    else
    {
        cout << "Negative weighted cycle detected\n";
    }

    return 0;
}
